src/logic/map_objects/tribes/ship.cc

```diff
@@ -883,6 +889,19 @@ WalkingDir Ship::get_scouting_direction() const {
 /// @note only called via player command
 void Ship::exp_construct_port(Game& game, const Coords& c) {
 	assert(expedition_);
+	// recheck ownership before setting the csite
+	Map* map = game.mutable_map();
+	MapRegion<Area<FCoords>> area(*map, Area<FCoords>(map->get_fcoords(c), 2));
+	do {
+		const PlayerNumber owner = area.location().field->get_owned_by();
+		if (owner != neutral() && owner != get_owner()->player_number()) {
+			// enemy player was faster conquering the spot
+			set_ship_state_and_notify(
+			   ShipStates::kExpeditionWaiting, NoteShip::Action::kDestinationChanged);
+			return;
+		}
+	} while (area.advance(*map));
+
```

Comment:
I really think `can_build_port_here()` should be used (I'm not sure if I got the syntax right though):

```suggestion
	// recheck ownership and immovables before setting the csite
	Map* map = game.mutable_map();
	if (get_owner()->player_number(), *map, !can_build_port_here(c)) {
		// enemy player was faster conquering the spot or player built something over it
		set_ship_state_and_notify(
		   ShipStates::kExpeditionWaiting, NoteShip::Action::kDestinationChanged);
		return;
	}

```

This way there's no need to maintain the code in 2 places, and checking buildings and flags is necessary, see 
[savegame](https://github.com/widelands/widelands/files/9974405/portspace_found_eliminated.zip)

Reply:
I am unsure whether a player would want to last its road here or just build the port.   
Furthermore this situation seems to be very constructed, i.e. why keeping the ship around and build something over the port?

Reply:
Yes, it's true that it's unlikely that anyone would do it intentionally. So the most likely use case would be that they click the build port button accidentally instead of a direction button or mistaking it for the window of another ship. But in these cases the buildings, roads and flags shouldn't be removed either.

But in my opinion, removing code duplication alone makes it worth the change.

Are you sure about buildings? I tested it, and the current implementation happily removes them.

Reply:
I am sure.  a building or constructionsite direct on the spot will not be removed.

Reply:
but I see the point about the code duplication and prevent unwanted actions. So i will making a change

Reply:
> I am sure. a building or constructionsite direct on the spot will not be removed.

I'd guess `force_csite()` does it. Please test it yourself. :)

Reply:
line 906 will prevent reaching the csite, but only for a building on the main spot of a port space. Teested it.
however I am working on a periodic check now as you sugested.   
Only it is hard to get the coords of the desired port space. will need some time and debugging

Reply:
Yes, sorry, that's true for the portspace location itself. What I tried was placing the building on one of the neighbouring nodes. That building is removed if the port is placed.

Thanks for trying the more involved feature too.